TopicOrPartitionUnknown exception thrown by KafkaStorage

The cause is that topic for kafka storage is created at runtime and kafka leader has not been elected or metadata not propagated to all brokers.

Also met this exception
[ERROR] [01/22/2015 12:03:04.689] [OneForOneStrategy] Failed to send messages after 3 tries.
kafka.common.FailedToSendMessageException: Failed to send messages after 3 tries.
        at kafka.producer.async.DefaultEventHandler.handle(DefaultEventHandler.scala:90)
        at kafka.producer.Producer.send(Producer.scala:76)
        at org.apache.gearpump.streaming.kafka.lib.KafkaStorage.append(KafkaStorage.scala:80)
        at org.apache.gearpump.streaming.kafka.lib.KafkaOffsetManager.filter(KafkaOffsetManager.scala:51)
        at org.apache.gearpump.streaming.kafka.KafkaSource.pullHelper$1(KafkaSource.scala:79)
        at org.apache.gearpump.streaming.kafka.KafkaSource.pull(KafkaSource.scala:89)
        at org.apache.gearpump.streaming.examples.stock.Reader.onNext(Reader.scala:30)
        at org.apache.gearpump.streaming.task.TaskActor.org$apache$gearpump$streaming$task$TaskActor$$doHandleMessage(TaskActor.scala:183)
        at org.apache.gearpump.streaming.task.TaskActor$$anonfun$stashAndHandleMessages$1.applyOrElse(TaskActor.scala:240)
        at akka.actor.Actor$class.aroundReceive(Actor.scala:465)
        at org.apache.gearpump.streaming.task.TaskActor.aroundReceive(TaskActor.scala:34)
        at akka.actor.ActorCell.receiveMessage(ActorCell.scala:516)
        at akka.actor.ActorCell.invoke(ActorCell.scala:487)
        at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:238)
        at akka.dispatch.Mailbox.run(Mailbox.scala:220)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
        at java.lang.Thread.run(Thread.java:745)

--- a/conf/kafka.conf
+++ b/conf/kafka.conf
@@ -18,6 +18,7 @@ kafka {
     producer.type = "sync"
     request.required.acks = "1"
     emit.batch.size = 100
+    fetch.metadata.timeout.ms = 10000
   }
 
   storage.replicas = 1

--- a/examples/kafka/src/test/scala/org/apache/gearpump/streaming/examples/kafka/KafkaStreamProducerSpec.scala
+++ b/examples/kafka/src/test/scala/org/apache/gearpump/streaming/examples/kafka/KafkaStreamProducerSpec.scala
@@ -101,6 +101,7 @@ class KafkaStreamProducerSpec extends PropSpec with Matchers with BeforeAndAfter
       KafkaConfig.SOCKET_TIMEOUT_MS -> "1000000",
       KafkaConfig.STORAGE_REPLICAS -> "1",
       KafkaConfig.ZOOKEEPER_CONNECT -> zookeeperConnect,
+      KafkaConfig.FETCH_METADATA_TIMEOUT_MS -> "3000",
       KafkaConfig.TIMESTAMP_FILTER_CLASS -> classOf[KafkaStreamProducerSpec.DummyFilter].getName,
       KafkaConfig.MESSAGE_DECODER_CLASS -> classOf[KafkaStreamProducerSpec.NoTimeStampDecoder].getName
     ).asJava)))

--- a/external/kafka/src/main/scala/org/apache/gearpump/streaming/kafka/lib/KafkaConfig.scala
+++ b/external/kafka/src/main/scala/org/apache/gearpump/streaming/kafka/lib/KafkaConfig.scala
@@ -50,6 +50,7 @@ object KafkaConfig {
   val PRODUCER_TYPE = "kafka.producer.producer.type"
   val REQUEST_REQUIRED_ACKS = "kafka.producer.request.required.acks"
   val PRODUCER_EMIT_BATCH_SIZE = "kafka.producer.emit.batch.size"
+  val FETCH_METADATA_TIMEOUT_MS = "kafka.producer.fetch.metadata.timeout.ms"
 
   // storage config
   val STORAGE_REPLICAS = "kafka.storage.replicas"
@@ -137,6 +138,10 @@ class KafkaConfig(config: Config) extends Serializable  {
 
   def getRequestRequiredAcks: String = {
     getString(REQUEST_REQUIRED_ACKS)
+  }
+
+  def getFetchMetadataTimeoutMS: Int = {
+    getInt(FETCH_METADATA_TIMEOUT_MS)
   }
 
   def getMetadataBrokerList: String = {

--- a/external/kafka/src/main/scala/org/apache/gearpump/streaming/kafka/lib/KafkaOffsetManager.scala
+++ b/external/kafka/src/main/scala/org/apache/gearpump/streaming/kafka/lib/KafkaOffsetManager.scala
@@ -36,6 +36,7 @@ object KafkaOffsetManager {
     val replicas = config.getStorageReplicas
     val zkClient = KafkaUtil.connectZookeeper(config)
     val topicExists = KafkaUtil.createTopic(zkClient, storageTopic, partitions = 1, replicas)
+    KafkaUtil.validateTopic(storageTopic, config.getMetadataBrokerList, config.getClientId, config.getFetchMetadataTimeoutMS)
     new KafkaOffsetManager(KafkaStorage(config, storageTopic, topicExists, topicAndPartition))
   }
 }

--- a/external/kafka/src/main/scala/org/apache/gearpump/streaming/kafka/lib/KafkaUtil.scala
+++ b/external/kafka/src/main/scala/org/apache/gearpump/streaming/kafka/lib/KafkaUtil.scala
@@ -21,6 +21,7 @@ package org.apache.gearpump.streaming.kafka.lib
 import java.util.Properties
 
 import kafka.admin.AdminUtils
+import kafka.client.ClientUtils
 import kafka.common.{TopicAndPartition, TopicExistsException}
 import kafka.producer.ProducerConfig
 import kafka.utils.{ZKStringSerializer, ZkUtils}
@@ -65,23 +66,26 @@ object KafkaUtil {
   /**
    *  create a new kafka topic
    *  return true if topic already exists, and false otherwise
-   *
-   *  Note: Do not call getBroker immediately after create topic as leader election and
-   *  metadata propagation takes time
    */
   def createTopic(zkClient: ZkClient, topic: String, partitions: Int, replicas: Int): Boolean = {
     try {
-      AdminUtils.createTopic(zkClient, topic, partitions, replicas)
-      LOG.info(s"created topic $topic")
-      false
-    } catch {
-      case tee: TopicExistsException =>
-        LOG.info(s"topic $topic already exists")
+      if (AdminUtils.topicExists(zkClient, topic)) {
+        LOG.info(s"topic $topic exists")
         true
+      } else {
+        AdminUtils.createTopic(zkClient, topic, partitions, replicas)
+        LOG.info(s"created topic $topic")
+        false
+      }
+    } catch {
       case e: Exception => throw e
     } finally {
       zkClient.close()
     }
+  }
+
+  def validateTopic(topic: String, brokerList: String, clientId: String, timeoutMs: Int) = {
+    ClientUtils.fetchTopicMetadata(Set(topic), ClientUtils.parseBrokerList(brokerList), clientId, timeoutMs)
   }
 
   def buildProducerConfig(config: KafkaConfig): ProducerConfig = {

--- a/external/kafka/src/test/scala/org/apache/gearpump/streaming/kafka/lib/KafkaConfigSpec.scala
+++ b/external/kafka/src/test/scala/org/apache/gearpump/streaming/kafka/lib/KafkaConfigSpec.scala
@@ -43,6 +43,7 @@ class KafkaConfigSpec extends PropSpec with Matchers with MockitoSugar {
     assert(Try(config.getProducerEmitBatchSize).isSuccess)
     assert(Try(config.getProducerType).isSuccess)
     assert(Try(config.getRequestRequiredAcks).isSuccess)
+    assert(Try(config.getFetchMetadataTimeoutMS).isSuccess)
     assert(Try(config.getMetadataBrokerList).isSuccess)
     assert(Try(config.getGrouperFactory).isSuccess)
     assert(Try(config.getStorageReplicas).isSuccess)

